File: python/binary_search_tree/validate_binary_search_tree.py
import math

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        return self.validate(root, -math.inf, math.inf)

    def validate(self,node: TreeNode, min_val: float, max_val: float) -> bool:
        if node is None:
            return True

        if node.val <= min_val or node.val >= max_val:
            return False

        return self.validate(node.left, min_val, node.val) and self.validate(node.right, node.val, max_val)

File: java/src/main/java/com/sanedge/interview/binary_search_tree/ValidateBinarySearchTree.java
package com.sanedge.interview.binary_search_tree;

import com.sanedge.interview.utils.TreeNode;

public class ValidateBinarySearchTree {

    public boolean isValidBST(TreeNode root) {
        return validate(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }

    public boolean validate(TreeNode node, int min, int max) {
        if (node == null) {
            return true;
        }

        if (node.val <= min || node.val >= max) {
            return false;
        }

        return (
            validate(node.left, min, node.val) &&
            validate(node.right, node.val, max)
        );
    }
}
